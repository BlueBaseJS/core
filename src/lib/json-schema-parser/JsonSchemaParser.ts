/**
 * Initial code take from https://github.com/TechniqueSoftware/react-json-schema
 */
import { MaybeArray } from '../../utils';
import React from 'react';
import isNil from 'lodash.isnil';
import isString from 'lodash.isstring';

export interface JsonComponentNode {
	component: string | React.ComponentType<any>;
	text?: string;
	props?: {
		key?: string;
		[key: string]: any;
	};
	children?: JsonComponentNode[];
	name?: string;
}

export type ResolveComponentFn = (node: JsonComponentNode) => (React.ComponentType<any> | null);

export class JsonSchemaParser {

	constructor(private getComponent: ResolveComponentFn = () => null) {
		//
	}

	parseSchema(schema: MaybeArray<JsonComponentNode>): MaybeArray<React.ReactElement<any>> {
		if (Array.isArray(schema)) {
			return this.parseSubSchemas(schema);
		}

		return this.createComponent(schema);
	}

	parseSubSchemas(subSchemas: JsonComponentNode[] = []): Array<React.ReactElement<any>> {
		const Components: Array<React.ReactElement<any>> = [];
		let index = 0;
		for (const subSchema of subSchemas) {
			subSchema.props = subSchema.props || {};
			subSchema.props.key = subSchema.props.key || String(index);
			const Component = this.parseSchema(subSchema) as React.ReactElement<any>;
			Components.push(Component);
			index++;
		}
		return Components;
	}

	createComponent(schema: JsonComponentNode): React.ReactElement<any> {
		const { text, props, component } = schema;

		if (isNil(component)) {
			throw Error(
				'Could not parse React JSON Schema. Reason: "component" property is required, but not given.'
			);
		}

		// // If schema.component is a React Component, return it
		// if (!isString(component) && React.isValidElement(component)) {
		// 	return component;
		// }

		const Component = this.resolveComponent(schema);
		const Children = text || this.resolveComponentChildren(schema);

		return React.createElement(Component, props, Children);
	}

	resolveComponent(node: JsonComponentNode): React.ComponentType<any> {

		// component is already a react component
		if (!isString(node.component) && node.component === Object(node.component)) {
			return node.component;
		}

		const Component = this.getComponent(node);

		if (Component) {
			return Component;
		}

		throw Error('Could not parse React JSON Schema. Reason: Could not resolve component.');
	}

	resolveComponentChildren({ children }: JsonComponentNode): MaybeArray<React.ReactElement<any>> | undefined {
		return !isNil(children) ? this.parseSchema(children) : undefined;
	}
}
