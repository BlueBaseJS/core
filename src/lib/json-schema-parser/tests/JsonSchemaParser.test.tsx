import { JsonComponentNode, JsonSchemaParser } from '../JsonSchemaParser';
import React from 'react';
import { Text } from 'react-native';
import renderer from 'react-test-renderer';

const Timeline = ({ children, ...rest }: any) => <Text {...rest}>Timeline: {children}</Text>;
const Post = ({ content, ...rest }: any) => <Text {...rest}>Post: {content}</Text>;

describe('JsonSchemaParser', () => {

	it('should parse schema with a single component', () => {

		const parser = new JsonSchemaParser();
		const result = parser.parseSchema({
			component: Timeline,
		});

		const tree = renderer.create(result as React.ReactElement<any>).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('should parse schema with a single component with props', () => {

		const parser = new JsonSchemaParser();
		const result = parser.parseSchema({
			component: Timeline,
			props: {
				style: {
					color: 'red'
				}
			}
		});

		const tree = renderer.create(result as React.ReactElement<any>).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('should parse schema with a single component with text', () => {

		const parser = new JsonSchemaParser();
		const result = parser.parseSchema({
			component: Timeline,
			text: 'Hello',
		});

		const tree = renderer.create(result as React.ReactElement<any>).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('should parse schema with a single component with children', () => {

		const parser = new JsonSchemaParser();
		const result = parser.parseSchema({
			component: Timeline,
			children: [{
				component: Post,
				props: { content: 'Hello' }
			}, {
				component: Post,
				props: { content: 'World' }
			}, {
				component: Post,
			}]
		});

		const tree = renderer.create(result as React.ReactElement<any>).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('should parse schema with custom component lookup', () => {

		const parser = new JsonSchemaParser((node: JsonComponentNode) => {
			return (node.component === 'Post') ? Post : null;
		});
		const result = parser.parseSchema({
			component: 'Post',
		});

		const tree = renderer.create(result as React.ReactElement<any>).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('should throw error with undefined component', () => {

		const parser = new JsonSchemaParser();
		const result = () => parser.parseSchema({
			component: undefined as any,
		});

		expect(result).toThrow();
	});

	it('should throw error with unknown component', () => {

		const parser = new JsonSchemaParser();
		const result = () => parser.parseSchema({
			component: 'Foo',
		});

		expect(result).toThrow();
	});

});