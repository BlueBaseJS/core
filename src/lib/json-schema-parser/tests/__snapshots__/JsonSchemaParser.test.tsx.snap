// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonSchemaParser should parse schema with a single component 1`] = `
<Text>
  Timeline: 
</Text>
`;

exports[`JsonSchemaParser should parse schema with a single component with children 1`] = `
<Text>
  Timeline: 
  <Text>
    Post: 
    Hello
  </Text>
  <Text>
    Post: 
    World
  </Text>
  <Text>
    Post: 
  </Text>
</Text>
`;

exports[`JsonSchemaParser should parse schema with a single component with props 1`] = `
<Text
  style={
    Object {
      "color": "red",
    }
  }
>
  Timeline: 
</Text>
`;

exports[`JsonSchemaParser should parse schema with a single component with text 1`] = `
<Text>
  Timeline: 
  Hello
</Text>
`;

exports[`JsonSchemaParser should parse schema with custom component lookup 1`] = `
<Text>
  Post: 
</Text>
`;
